src/TreeDataGrid.tsx

```diff
@@ -0,0 +1,428 @@
+  const leftKey = isRtl ? 'ArrowRight' : 'ArrowLeft';
+  const rightKey = isRtl ? 'ArrowLeft' : 'ArrowRight';
+  const toggleGroupLatest = useLatestFunc(toggleGroup);
+
+  const { columns, groupBy } = useMemo(() => {
+    const columns = [...rawColumns].sort(({ key: aKey }, { key: bKey }) => {
+      // Sort select column first:
+      if (aKey === SELECT_COLUMN_KEY) return -1;
+      if (bKey === SELECT_COLUMN_KEY) return 1;
+
+      // Sort grouped columns second, following the groupBy order:
+      if (rawGroupBy.includes(aKey)) {
+        if (rawGroupBy.includes(bKey)) {
+          return rawGroupBy.indexOf(aKey) - rawGroupBy.indexOf(bKey);
+        }
+        return -1;
+      }
+      if (rawGroupBy.includes(bKey)) return 1;
+
+      // Sort other columns last:
+      return 0;
+    });
+
+    const groupBy: string[] = [];
+    for (const [index, column] of columns.entries()) {
+      if (rawGroupBy.includes(column.key)) {
+        groupBy.push(column.key);
+        columns[index] = {
+          ...column,
+          frozen: true,
+          renderCell: () => null,
+          renderGroupCell: column.renderGroupCell ?? renderToggleGroup,
+          editable: false
+        };
+      }
+    }
+
+    return { columns, groupBy };
+  }, [rawColumns, rawGroupBy]);
+
+  const [groupedRows, rowsCount] = useMemo(() => {
+    if (groupBy.length === 0) return [undefined, rawRows.length];
+
+    const groupRows = (
+      rows: readonly R[],
+      [groupByKey, ...remainingGroupByKeys]: readonly string[],
+      startRowIndex: number
+    ): [Readonly<GroupByDictionary<R>>, number] => {
+      let groupRowsCount = 0;
+      const groups: GroupByDictionary<R> = {};
+      for (const [key, childRows] of Object.entries(rowGrouper(rows, groupByKey))) {
+        // Recursively group each parent group
+        const [childGroups, childRowsCount] =
+          remainingGroupByKeys.length === 0
+            ? [childRows, childRows.length]
+            : groupRows(childRows, remainingGroupByKeys, startRowIndex + groupRowsCount + 1); // 1 for parent row
+        groups[key] = { childRows, childGroups, startRowIndex: startRowIndex + groupRowsCount };
+        groupRowsCount += childRowsCount + 1; // 1 for parent row
+      }
+
+      return [groups, groupRowsCount];
+    };
+
+    return groupRows(rawRows, groupBy, 0);
+  }, [groupBy, rowGrouper, rawRows]);
+
+  const [rows, isGroupRow] = useMemo((): [
+    ReadonlyArray<R | GroupRow<R>>,
+    (row: R | GroupRow<R>) => row is GroupRow<R>
+  ] => {
+    const allGroupRows = new Set<unknown>();
+    if (!groupedRows) return [rawRows, isGroupRow];
+
+    const flattenedRows: Array<R | GroupRow<R>> = [];
+    const expandGroup = (
+      rows: GroupByDictionary<R> | readonly R[],
+      parentId: string | undefined,
+      level: number
+    ): void => {
+      if (isReadonlyArray(rows)) {
+        flattenedRows.push(...rows);
+        return;
+      }
+      Object.keys(rows).forEach((groupKey, posInSet, keys) => {
```

Comment:
what do you think about using `for...of` here?

Reply:
Here we need access to the array ( `keys`) itself